src/db/tables/emailAddresses.ts

```diff
@@ -43,22 +44,32 @@ async function addSubscriberUnverifiedEmailHash(
   email: string,
 ) {
   const lowerCaseEmail = email.toLowerCase();
-  const res = await knex.transaction((trx) => {
-    return trx("email_addresses")
+
+  const res = await knex.transaction(async (trx) => {
+    await trx.raw("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE");
+    return await trx("email_addresses")
       .forUpdate()
-      .select({
-        subscriber_id: user.id,
-      })
-      .insert({
-        subscriber_id: user.id,
-        email: lowerCaseEmail,
-        sha1: getSha1(lowerCaseEmail),
-        verification_token: uuidv4(),
-        verified: false,
-      })
-      .returning("*");
+      .select()
+      .where("subscriber_id", user.id)
+      .then(async (rows) => {
+        if (rows.length < CONST_MAX_NUM_ADDRESSES - 1)
+          return trx("email_addresses")
+            .insert({
+              subscriber_id: user.id,
+              email: lowerCaseEmail,
+              sha1: getSha1(lowerCaseEmail),
+              verification_token: uuidv4(),
+              verified: false,
+            })
+            .returning("*");
+      });
   });
-  return res[0];
+
+  if (res) {
+    return res[0];
+  } else {
+    return false;
```

Comment:
Looks like you might have to change this for type check

Reply:
Ah thanks I guess it didn't break the quick check-in hook lint run... I'll fix this up. I think the caller should probably handle it being possibly `false` (or maybe `undefined`), or we should have it throw.

Reply:
I switched it to throw an exception.